sympy/core/operations.py

```diff
@@ -504,3 +510,149 @@ def args(self):
+       See https://github.com/sympy/sympy/pull/19463.
+
+    Explanation
+    ===========
+
+    If arguments of different kind are passed, their ``_op_priority`` are compared to
+    choose the type with the highest priority. If this is ambiguous, priority relations
+    which are registered by ``register_priority`` method are referred to.
+    Once a type is selected, handler method of the type, which is a static method, is used
+    to perform the operation.
+
+    Priority registeration is unordered. You cannot make ``A*B`` and ``B*A`` use different
+    handler method. Every logic dealing with the order of arguments must be implemented in
+    the handler method.
+
+    Examples
+    ========
+
+    >>> from sympy import Add, Expr, Symbol
+    >>> from sympy.core.add import add
+
+    >>> class NewExpr(Expr):
+    ...     @staticmethod
+    ...     def _add_handler(*args, **kwargs):
+    ...         return NewAdd(*args, **kwargs)
+    >>> class NewAdd(NewExpr, Add):
+    ...     pass
+
+    >>> @add.register_priority(Expr, NewExpr)
+    ... @add.register_priority(NewExpr, NewExpr)
+    ... def _(_1, _2):
+    ...     return NewExpr
+
+    >>> a, b = Symbol('a'), NewExpr()
+    >>> add(a, b) == NewAdd(a, b)
+    True
+
+    Registered priority selector can return any type.
+
+    >>> class Foo(Expr):
+    ...     @staticmethod
+    ...     def _add_handler(*args, **kwargs):
+    ...         raise NotImplementedError
+    >>> class Bar(Expr):
+    ...     @staticmethod
+    ...     def _add_handler(*args, **kwargs):
+    ...         raise NotImplementedError
+    >>> class FooBar(Expr):
+    ...     @staticmethod
+    ...     def _add_handler(*args, **kwargs):
+    ...         return NewAdd(*args, **kwargs)
+
+    >>> @add.register_priority(Foo, Bar)
+    ... def _(_1, _2):
+    ...     return FooBar
+
+    >>> add(Foo(), Bar()) == NewAdd(Foo(), Bar())
+    True
+
+    """
+    def __init__(self, name, doc=None):
+        self.name = name
+        self.doc = doc
+        self.handlermethod = "_%s_handler" % name
+        self.handlergetter = attrgetter(self.handlermethod)
+        self._dispatcher = Dispatcher(name)
+
+    def __repr__(self):
+        return "<dispatched %s>" % self.name
+
+    def register_priority(self, cls1, cls2, on_ambiguity=ambiguity_register_error):
+        """
+        Register the priority between two types, in both straight and reversed order.
+        Registered function must return the class which has the handler that is to be used.
+        """
+        def _(func):
+            self._dispatcher.add((cls1, cls2), func, on_ambiguity=on_ambiguity)
+            self._dispatcher.add((cls2, cls1), func, on_ambiguity=on_ambiguity)
+            return func
+        return _
+
+    @cacheit
+    def __call__(self, *args, **kwargs):
+        if kwargs.get('sympify', True):
+            args = tuple([_sympify(arg) for arg in args])
+        types = tuple([type(arg) for arg in args])
```

Comment:
I think that this method should always sympify the inputs unless there is a significant slowdown from sympifying already sympified args. Internal calculations should always use sympified objects so we should benchmark with those if the cost of sympification is significant.

Does the ordering of types matter here?

The first thing `dispatch` does is turn this into a `set` which means that the above creates a list only to convert that to a tuple which is then immediately converted into a set. I think I'd change this to:
```
args = tuple(map(_sympify, args))
types = frozenset(map(type, args))
```
and then change `dispatch` so it takes `types` rather than `*types`. The `cacheit` on `dispatch` is much more likely to get cache hits for a `frozenset` of types than a tuple of all of the types.

I think @eric-wieser suggested that it might be better to cache the pairwise dispatch instead which might be better. I'm not sure but I think that maybe multiple dispatch already has an internal cache for pairwise dispatch.

Apart from `evaluate` what possible `kwargs` are there? It's best to avoid `**kwargs` if there isn't any particular reason for it.

Reply:
There are cases in core where we can guarantee that arguments are already sympified (e.g. `Add` in `_eval_expand_mul` - where `Add` will be changed to `add`, of course). I introduced `sympify` kwarg only for these cases. Otherwise, sympifying will be done by default.  
What problem will `if kwargs.get('sympify', True):` condition raise, do you expect?